Revamp onboarding docs

neverett opened the pull request:
My goal is to create an updated set of onboarding docs that makes it easier for new users to get started with Flyte. These docs will have the following structure:

* **Introduction to Flyte:** Getting Started docs landing page
* **Quickstart guide:** Quick intro to creating and running a workflow
* **Getting started with Flyte development:** New section to hold the pages below
  * **Index page**
  * **Installing development tools**
  * **Creating a Flyte Project**
  * **Flyte project components**
  * **Running Flyte workflows and tasks locally**


### TODO

#### Before merging this PR

- [x] Merge  https://github.com/flyteorg/static-resources/pull/31
- [x] Update image links in running_a_workflow_locally.md to point to main branch of static-resources repo
- [x] Merge https://github.com/flyteorg/flytekit-python-template/pull/43 to add a new ImageSpec-based flytekit template
- [x] Merge https://github.com/flyteorg/flytekit/pull/2041 to update flytekit to use the new basic ImageSpec template by default for `pyflyte init`
- [ ] Add redirect for "Creating a Flyte project" (old file: creating_flyte_project.md, new file: creating_a_flyte_project.md)
- [ ] Add redirect for "Visualizing artifacts"

#### After merging this PR

- [ ] Close https://github.com/flyteorg/flyte/issues/3243
- [ ] Close https://github.com/flyteorg/flyte/issues/3520
- [ ] Merge https://github.com/flyteorg/flyte/pull/4548

neverett commented on docs/index.md:
```diff
@@ -4,387 +4,50 @@ jupytext:
 # override the toc-determined page navigation order
-next-page: getting_started/flyte_fundamentals
-next-page-title: Flyte Fundamentals
+next-page: getting_started/quickstart_guide
```
The "next" link is currently not rendering correctly (points to the wrong place), but I'm not sure why.

cosmicBboy replied:
We won't actually need this after monodocs is merged. Typically the page order is determined by the `toctree` directive. The `next-page`, `prev-page` thing is custom behavior that was meant to glue together the 5 separate docs sites.

neverett commented on docs/getting_started/quickstart_guide.md:
```diff
@@ -0,0 +1,69 @@
+
+# override the toc-determined page navigation order
+prev-page: index
+prev-page-title: Introduction to Flyte
```
The "prev" link is currently not working currently, but I'm not sure why.

neverett commented on docs/getting_started/quickstart_guide.md:
```diff
@@ -0,0 +1,69 @@
+
+## Prerequisites
+
+* Install Python
```
What version?

neverett replied:
Python 3.x+

neverett commented on docs/index.md:
`````diff
@@ -4,387 +4,50 @@ jupytext:
+* The **control plane** is part of the Flyte backend that is configured by platform engineers or others tasked with setting up computing infrastructure. It consists of FlyteAdmin, which serves as the main Flyte API to process requests from clients in the user plane. The control plane sends workflow execution requests to the data plane for execution, and stores information such as current and past ...
+* The **data plane** is another part of the Flyte backend that contains FlytePropeller, the core engine of Flyte that executes workflows. FlytePropeller is designed as a [Kubernetes Controller](https://kubernetes.io/docs/concepts/architecture/controller/). The data plane sends status events back to the control plane so that information can be stored and surfaced to end users.
 
-````{dropdown} Want to try Flyte on the browser?
`````
I removed this link since it takes people out of the docs and puts them into an entirely different onboarding path, which seems confusing. I do think the hosted sandbox is helpful, but it seems better to link to it from the Flyte website (where it's already linked).

neverett commented on docs/getting_started/creating_a_flyte_project.md:
```diff
@@ -0,0 +1,79 @@
+    format_name: myst
+---
+
+# Creating a Flyte project
```
Currently, "Flyte project" refers to two completely different things, which is very confusing. Ideally, we would rename one, but I'm not sure how feasible that is.

EngHabu replied:
I'm torn between introducing git... etc. here vs jumping right to running things...

neverett commented on docs/getting_started/creating_a_flyte_project.md:
```diff
@@ -0,0 +1,79 @@
+
+Next, initialize your Flyte project. The [flytekit-python-template GitHub repository](https://github.com/flyteorg/flytekit-python-template) contains Flyte project templates with sample code that you can run as is or modify to suit your needs.
+
+In this example, we will initialize the [simple-example project template](https://github.com/flyteorg/flytekit-python-template/tree/main/simple-example).
```
For now, I used the simple example project template, but it would be nice to use a classification example for the ML audience. There is a wine classification project template, which is used in the Union docs, but in my opinion, that's not ideal—a lot of people don't drink alcohol (for a variety of reasons), so the example reads a little charged and specific. It would be great to have a more neutral classification example; the classic one uses the [Iris dataset](https://archive.ics.uci.edu/dataset/53/iris), so that might be a good one to include.

neverett commented on docs/getting_started/running_a_workflow_locally.md:
```diff
@@ -0,0 +1,94 @@
+
+# Running a workflow locally
+
+## Running a workflow in a local Python environment
```
Depending on what we want to emphasize in this article, we should do one of two things:
* Include a note about the limitations of this approach in this section.
* Remove this section, make the entire article about running a workflow in a local cluster, and add a note about how you can run in a local Python environment as well, but that is not a recommended approach.

neverett commented on docs/getting_started/flyte_project_components.md:
```diff
@@ -0,0 +1,101 @@
+
+(getting_started_dockerfile)=
+
+### Dockerfile
```
TODO: Decide if we want to mention ImageSpec here, and if so, if we also want to actively discourage people from customizing Dockerfiles.

EngHabu replied:
I think we should remove this section and ONLY mention ImageSpec... maybe with a link at the end `for advanced image configurations, read on <building with Dockerfile> instead`

neverett commented on docs/getting_started/creating_a_flyte_project.md:
```diff
@@ -0,0 +1,79 @@
+
+### Version your Flyte project code with git
+
+Versioning your Flyte project code allows you to register the workflow it contains to a Flyte cluster and enables workflow versioning.
```
Specifically, initializing a git repo and creating at least one commit is necessary for containerizing your project.

neverett commented on docs/getting_started/creating_a_flyte_project.md:
```diff
@@ -0,0 +1,79 @@
+    format_name: myst
+---
+
+# Creating a Flyte project
```
If you run the workflow in the `my_project` Flyte project in the demo cluster by running `pyflyte run --remote` (instead of packaging and registering the project), FlyteConsole will not show the name of your project—it says "Flytesnacks" instead, since that's the project name associated with the demo cluster. We should probably mention that in either this article or "Running a workflow locally", or both, to reduce confusion.

EngHabu replied:
`pyflyte run --remote -p <my project> -d development` should do the trick.. have you tried that?

EngHabu commented on docs/index.md:
```diff
@@ -4,387 +4,50 @@ jupytext:
-This introduction provides a quick overview of how to get Flyte up and running
-on your local machine.
+Flyte is made up of a user plane, control plane, and data plane.
+* The **user plane** contains the elements you need to develop the code that will implement your application's directed acyclic graph (DAG). These elements are FlyteKit, the FlyteConsole, and Flytectl. Data scientists and machine learning engineers primarily work in the user plane.
```
I would move FlyteConsole to control plane

EngHabu commented on docs/getting_started/installing_development_tools.md:
```diff
@@ -0,0 +1,94 @@
+
+## Install Python
+
+Python versions 3.8x - 3.10x are supported. [TK - is this accurate?]
```
Just want to leave a note to not forget to update the question when you have an answer :)

neverett replied:
Python 3.x

EngHabu commented on docs/getting_started/installing_development_tools.md:
```diff
@@ -0,0 +1,94 @@
+
+# Installing development tools
+
+To create and run workflows in Flyte, you must install Python, flytekit, flytectl, and Docker. Installing conda or another Python virtual environment manager is optional, but strongly recommended.
```
I think our steps for onboarding should be just `pip install flytekit` any more tools and it gets too much... flytectl and docker are needed when we get to running on the demo cluster, right? can these steps be postponed until then?

EngHabu commented on docs/getting_started/running_a_workflow_locally.md:
```diff
@@ -0,0 +1,94 @@
+    format_name: myst
+---
+
+# Running a workflow locally
```
I think this should come right after creating a project.

EngHabu commented on docs/getting_started/running_a_workflow_locally.md:
````diff
@@ -0,0 +1,94 @@
+```
+2. Run the workflow on the Flyte cluster with `pyflyte run`:
+```{prompt} bash $
+pyflyte run --remote example.py wf
````
This assumes `flytectl demo start` has run... I'm not sure this is clear here... or did you mean to just use `pyflyte run example.py wf` ? 
I think emphasizing the fact that the same code runs in a regular python environment to allow iterating super quick and then jumping to `pyflyte run --remote` to run on a local/remote cluster (potentially with more resources) is the story to tell...

EngHabu commented on docs/getting_started/flyte_fundamentals.md:
```diff
@@ -52,7 +50,6 @@ cluster, see the {ref}`Deployment Guide <deployment>`.
-creating_flyte_project
 package_register
 run_schedule
 visualizing_artifacts
```
I think we should stop calling it artifacts because we are about to introduce an artifacts service... might be a good time to change the name here...

EngHabu reviewed (commented):
Thank you Nikki for this revamp! I left a few comments but we can discuss in the meeting today.

Shouldn't the "Next" be the quick start guide?

<img width="861" alt="image" src="https://github.com/flyteorg/flytesnacks/assets/27159/2880294b-76b2-48ea-a0e8-057b23dcdb48">

neverett commented on docs/getting_started/flyte_project_components.md:
`````diff
@@ -0,0 +1,87 @@
+
+````{dropdown} See requirements.txt
+
+```{rli} https://raw.githubusercontent.com/flyteorg/flytekit-python-template/main/simple-example/%7B%7Bcookiecutter.project_name%7D%7D/requirements.txt
`````
This link will need to be updated after https://github.com/flyteorg/flytekit-python-template/pull/43 is merged

brndnblck commented on docs/getting_started/flyte_project_components.md:
```diff
@@ -0,0 +1,90 @@
+
+(getting_started_workflow_code)=
+
+## Workflow code
```
Nitpick, but should this be title case?

```suggestion
## Workflow Code
```

neverett replied:
In general, Peeter and I decided on using sentence case in docs, since it makes proper nouns (like product and feature names) easier to distinguish and eliminates guessing at whether words like "a", "an", "the", etc, should be capitalized.

EngHabu commented on docs/getting_started/running_a_workflow_locally.md:
````diff
@@ -0,0 +1,168 @@
+    print(hello_world_wf())
+```
+
+Your code would become even more verbose if you wanted to pass arguments to the workflow:
+```python
+if __name__ == "__main__":
+    from argparse import ArgumentParser
+
+    parser = ArgumentParser()
+    parser.add_argument("--name", type=str)
+
+    args = parser.parse_args()
+    print(hello_world_wf(name=args.name))
+```
+:::
````
This shouldn't be needed. `pyflyte run file.py hello_world_wf` should ask you to enter the `--name` already...

neverett replied:
Understood, the purpose of this note is to explain why we recommend using `pyflyte run` rather than `python example.py` and is pulled from the [existing docs](https://docs.flyte.org/projects/cookbook/en/latest/index.html), but I can remove it if it seems confusing.

EngHabu commented on docs/getting_started/flyte_project_components.md:
````diff
@@ -0,0 +1,90 @@
+
+A Flyte project is a directory containing task and workflow code, internal Python source code, configuration files, and other artifacts needed to package up your code so that it can be run on a Flyte cluster.
+
+## Directory structure
+
+If you examine the project you created with `pyflyte init` in {doc}`"Creating a Flyte project" <creating_a_flyte_project>`, you'll see the following directory structure:
+
+```{code-block} bash
+my_project
+├── LICENSE
+├── README.md
+├── requirements.txt  # Python dependencies
+└── workflows
+    ├── __init__.py
+    └── example.py    # Example Flyte workflow code
+```
+
````
Discussed Offline: We will simplify the flytekit template repo (at least the one to use in this getting started guide) to just be a single file. While maintaining a separate template that looks like this ☝🏽

EngHabu reviewed (commented):
Great progress! Thank you all your work
 
https://flyte--1266.org.readthedocs.build/projects/cookbook/en/1266/index.html the "Next" here still takes you to "User Guide"

neverett commented:
**Update:** I added a "Hello, world!" flytekit template to [this PR](https://github.com/flyteorg/flytekit-python-template/pull/43), and used the "Hello, world!" template in the [Quickstart guide](https://flyte--1266.org.readthedocs.build/projects/cookbook/en/1266/getting_started/quickstart_guide.html), since that seemed like the best place for a pared-down example. (As a bonus, it made the quickstart guide shorter.)

The rest of the articles in the "Getting started with workflow development" section are intended to guide users beyond the simple example and teach them to use Flyte in a productionizable way, so they demonstrate project creation with the new default ImageSpec template.

neverett commented:
@cosmicBboy the "Next" link on the [Getting Started index page](https://flyte--1266.org.readthedocs.build/projects/cookbook/en/1266/index.html) and the "Prev" link on the [Quickstart guide ](https://flyte--1266.org.readthedocs.build/projects/cookbook/en/1266/getting_started/quickstart_guide.html)are wrong, but I'm not sure why, since I updated the index page TOC, and added next/prev links in the front matter for both pages ([index](https://github.com/flyteorg/flytesnacks/blob/nikki-onboarding-docs-revamp/docs/index.md), [quickstart](https://github.com/flyteorg/flytesnacks/blob/nikki-onboarding-docs-revamp/docs/getting_started/quickstart_guide.md)). Is there another place I need to update the TOC?

cosmicBboy commented:
> @cosmicBboy the "Next" link on the [Getting Started index page](https://flyte--1266.org.readthedocs.build/projects/cookbook/en/1266/index.html) and the "Prev" link on the [Quickstart guide ](https://flyte--1266.org.readthedocs.build/projects/cookbook/en/1266/getting_started/quickstart_guide.html)are wrong, but I'm not sure why, since I updated the index page TOC, and added next/prev links in the front matter for both pages ([index](https://github.com/flyteorg/flytesnacks/blob/nikki-onboarding-docs-revamp/docs/index.md), [quickstart](https://github.com/flyteorg/flytesnacks/blob/nikki-onboarding-docs-revamp/docs/getting_started/quickstart_guide.md)). Is there another place I need to update the TOC?

Hi @neverett, we don't need to worry about this after we switch over to monodocs, so I wouldn't worry about this too much
